packages/ra-core/src/controller/RecordContext.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import { createContext, useContext, ReactNode } from 'react';
+import pick from 'lodash/pick';
+import { Record } from '../types';
+
+/**
+ * Context to store the result of the useRecord() hook.
+ *
+ * Use the useRecordContext() hook to read the context. That's what the Edit and Show components do in react-admn.
+ *
+ * @example
+ *
+ * import { useEditController, EditContext } from 'ra-core';
+ *
+ * const Edit = props => {
+ *     const controllerProps = useEditController(props);
+ *     return (
+ *         <RecordContextProvider value={controllerProps}>
+ *             ...
+ *         </RecordContextProvider>
+ *     );
+ * };
+ */
+export const RecordContext = createContext<Record | Omit<Record, 'id'>>(
+    undefined
+);
+
+export const RecordContextProvider = ({
+    children,
+    value,
+}: RecordContextOptions) => (
+    <RecordContext.Provider value={value}>{children}</RecordContext.Provider>
+);
+RecordContext.displayName = 'RecordContext';
+
+export const usePickRecordContext = <
+    RecordType extends Record | Omit<Record, 'id'> = Record
+>(
+    context: RecordType
+) => {
+    return pick(context, ['record']);
```

Comment:
I'm thinking: every time something changes in a context (e.g. the selected ids in the list), the pick command will return a new instance of the record, re-rendering everything. 

I think it's not a premature optimization to memoize the picking.